### ISSUE TITLE
Jars should include directories

### ISSUE THREAD

**Issue by tballard** [com-lihaoyi/mill#2136]
Mill builds jars without entries for directories, but directories are cheap and allow searching for resources and are otherwise harmless and thus should be included. (Evidence suggests that maven includes them by default, supporting the claim of being benign)

**Comment by lolgab** [comment]
Can you provide an example?

**Comment by lefou** [comment]
Yeah, I created jars with so many build tools and even coded that part multiple times from scratch for various build tools. It could appear as nitpicking to argue for or against additionally explicit directory entries. I nowadays think, it's probably easier to read and probably also easier to process, when these entries are present. And when you told me, that [Spring can't find resources properly](https://gitter.im/lihaoyi/mill?at=637cfaa6426f9311ffb83a43) when directory entries are missing, you had me. It's probably a poor implementation on their side, but nevertheless, it inconvenient and avoidable.

Only thing I am very opinionated about is, that we keep the `META-INF/MANIFEST.MF` entry as first and before the `META-INF` directory entry. This is more in line with the early specs and also makes it much easier to look into jars, e.g. with `less my.jar`, which especially helps if you work with OSGi or Manifests in general a lot.

**Comment by tballard** [comment]
I think the directories may be supposed to precede their contents, too

**Comment by lefou** [comment]
> I think the directories may be supposed to precede their contents, too

Sure. That is also my assumption/understanding. That's why I pointed to the special case with `META-INF/`.

**Comment by lefou** [comment]
@tballard Do you mind to give PR #2138 a try?

**Comment by tballard** [comment]
Yeah I will. I'm writing some code to repack the current one, adding the required directories, so I'll be all practiced up.

**Comment by tballard** [comment]
Hmm, I thought I was implementing this, although I like your solution better. However, you will note that if you try to add an entry like META-INF/SOMETHING_ELSE you get a collision:
1 targets failed
reds.jar java.util.zip.ZipException: duplicate entry: META-INF/
    java.base/java.util.zip.ZipOutputStream.putNextEntry(ZipOutputStream.java:241)

**Comment by lefou** [comment]
> Hmm, I thought I was implementing this, although I like your solution better.

Oh, I though you will start it once you are on holiday, but I didn't realized that this means "now". 

> However, you will note that if you try to add an entry like META-INF/SOMETHING_ELSE you get a collision: 1 targets failed reds.jar java.util.zip.ZipException: duplicate entry: META-INF/ java.base/java.util.zip.ZipOutputStream.putNextEntry(ZipOutputStream.java:241)

Oh, this is an oversight, as the special case `META-INF/` needs to be explicitly added to the `seen` set. Thanks for testing and pointing this out. I'll push a fix.

**Comment by tballard** [comment]
Yeah, I had that bug first;-)
